File: src/VersionUp/VersionHandlers/AssemblyInfoVersionHandler.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

namespace VersionUp.VersionHandlers
{
	/// <summary>
	/// Handles parsing and updating version attributes in C#, VB, and F# AssemblyInfo files.
	/// </summary>
	public class AssemblyInfoVersionHandler : IVersionFileHandler
	{
	    /// <summary>
	    /// Matches <c>AssemblyVersion</c>, <c>AssemblyFileVersion</c>, and
	    /// <c>AssemblyInformationalVersion</c> attribute declarations in C#, VB, and F# syntax.
	    /// </summary>
	    private static readonly Regex VersionRegex = new Regex(
	        @"(?:\[<assembly:\s*|\[assembly:\s*|<Assembly:\s*)(AssemblyVersion|AssemblyFileVersion|AssemblyInformationalVersion)\(\""([^\""]+)\""\)(?:\]>|\]|>)",
	        RegexOptions.Compiled | RegexOptions.IgnoreCase);

	    /// <inheritdoc />
	    public bool CanHandle(string filePath)
	    {
	        if (string.IsNullOrEmpty(filePath))
	        {
	            return false;
	        }

	        string fileName = Path.GetFileNameWithoutExtension(filePath).ToLowerInvariant();
	        string extension = Path.GetExtension(filePath).ToLowerInvariant();

	        bool isAssemblyInfo = fileName.Equals("assemblyinfo", StringComparison.OrdinalIgnoreCase);
	        bool isSupportedExtension = extension == ".cs" || extension == ".vb" || extension == ".fs";

	        return isAssemblyInfo && isSupportedExtension;
	    }

	    /// <inheritdoc />
	    public string? GetVersion(string fileContent)
	    {
	        if (string.IsNullOrWhiteSpace(fileContent))
	        {
	            return null;
	        }

	        MatchCollection matches = VersionRegex.Matches(fileContent);

	        foreach (Match match in matches)
	        {
	            if (match.Groups[1].Value.Equals("AssemblyVersion", StringComparison.OrdinalIgnoreCase))
	            {
	                string value = match.Groups[2].Value;

	                return value;
	            }
	        }

	        foreach (Match match in matches)
	        {
	            if (match.Groups[1].Value.Equals("AssemblyFileVersion", StringComparison.OrdinalIgnoreCase))
	            {
	                string value = match.Groups[2].Value;

	                return value;
	            }
	        }

	        if (matches.Count > 0)
	        {
	            string value = matches[0].Groups[2].Value;

	            return value;
	        }

	        return null;
	    }

	    /// <inheritdoc />
	    public string UpdateVersion(string fileContent, string newVersion)
	    {
	        if (string.IsNullOrWhiteSpace(fileContent))
	        {
	            return fileContent;
	        }

	        string result = VersionRegex.Replace(fileContent, match =>
	        {
	            string attributeName = match.Groups[1].Value;
	            string rawMatch = match.Value;
	            string rawValue = match.Groups[2].Value;

	            int index = rawMatch.IndexOf(rawValue, StringComparison.Ordinal);

	            if (index >= 0)
	            {
	                string replaced = rawMatch.Substring(0, index) + newVersion + rawMatch.Substring(index + rawValue.Length);

	                return replaced;
	            }

	            return rawMatch;
	        });

	        return result;
	    }
	}
}

File: vscode-extension/src/handlers/AssemblyInfoVersionHandler.ts
import * as path from 'path';
import { IVersionFileHandler } from './IVersionFileHandler';

/**
 * Handles parsing and updating version attributes in C#, VB, and F# AssemblyInfo files.
 */
export class AssemblyInfoVersionHandler implements IVersionFileHandler {
    private static readonly versionRegex = /(?:\[<assembly:\s*|\[assembly:\s*|<Assembly:\s*)(AssemblyVersion|AssemblyFileVersion|AssemblyInformationalVersion)\("([^"]+)"\)(?:\]>|\]|>)/gi;

    public canHandle(filePath: string): boolean {
        if (!filePath) {
            return false;
        }

        const fileName = path.basename(filePath).toLowerCase();
        const ext = path.extname(filePath).toLowerCase();

        const isAssemblyInfo = fileName.startsWith('assemblyinfo');
        const isSupportedExtension = ext === '.cs' || ext === '.vb' || ext === '.fs';

        return isAssemblyInfo && isSupportedExtension;
    }

    public getVersion(fileContent: string): string | null {
        if (!fileContent || !fileContent.trim()) {
            return null;
        }

        AssemblyInfoVersionHandler.versionRegex.lastIndex = 0;
        const matches = [...fileContent.matchAll(AssemblyInfoVersionHandler.versionRegex)];

        for (const match of matches) {
            if (match[1].toLowerCase() === 'assemblyversion') {
                return match[2];
            }
        }

        for (const match of matches) {
            if (match[1].toLowerCase() === 'assemblyfileversion') {
                return match[2];
            }
        }

        if (matches.length > 0) {
            return matches[0][2];
        }

        return null;
    }

    public updateVersion(fileContent: string, newVersion: string): string {
        if (!fileContent || !fileContent.trim()) {
            return fileContent;
        }

        AssemblyInfoVersionHandler.versionRegex.lastIndex = 0;
        return fileContent.replace(AssemblyInfoVersionHandler.versionRegex, (match, _attributeName, rawValue) => {
            const index = match.indexOf(rawValue);
            if (index >= 0) {
                return match.slice(0, index) + newVersion + match.slice(index + rawValue.length);
            }
            return match;
        });
    }
}
